base/linalg/dense.jl

```diff
@@ -265,29 +266,50 @@ julia> diag(A,1)
 diag(A::AbstractMatrix, k::Integer=0) = A[diagind(A,k)]
 
 """
-    diagm(v, k::Integer=0)
+    diagm(kv::Pair{<:Integer,<:AbstractVector}...)
+
+Construct a square matrix from `Pair`s of diagonals and vectors.
+Vector `kv.second` will be placed on the `kv.first` diagonal.
 
-Construct a matrix by placing `v` on the `k`th diagonal. This constructs a full matrix; if
-you want a storage-efficient version with fast arithmetic, use [`Diagonal`](@ref) instead.
+See also: [`spdiagm`](@ref), [`Diagonal`](@ref)
 
 # Examples
 ```jldoctest
-julia> diagm([1,2,3],1)
+julia> diagm(1 => [1,2,3])
 4×4 Array{Int64,2}:
  0  1  0  0
  0  0  2  0
  0  0  0  3
  0  0  0  0
+
+julia> diagm(1 => [1,2,3], -1 => [4,5])
+4×4 Array{Int64,2}:
+ 0  1  0  0
+ 4  0  2  0
+ 0  5  0  3
+ 0  0  0  0
 ```
 """
-function diagm(v::AbstractVector{T}, k::Integer=0) where T
-    n = length(v) + abs(k)
-    A = zeros(T,n,n)
-    A[diagind(A,k)] = v
-    A
+function diagm(kv::Pair{<:Integer,<:AbstractVector}...)
+    A = diagm_container(kv...)
+    for p in kv
+        inds = diagind(A, p.first)
+        for (i, val) in enumerate(p.second)
+            A[inds[i]] += val
```

Comment:
Why `+=` rather than `=`?

Reply:
To be consistent with `spdiagm`, since `combine` in `sparse` defaults to `+`. See https://github.com/JuliaLang/julia/pull/24047#discussion_r144005196 . Do you have another opinion about this?

Reply:
Ah, do IIUC that its purpose is e.g. `[sp]diagm(0 => [1, 2, 3], 0 => [4, 5, 6])` yielding `[sp]diagm(0 => [5, 7, 9])`? If so, cheers, and thanks! :)

Reply:
Yes, I think it is the most natural thing to do.